runtime/runtime/src/lib.rs

```diff
@@ -291,127 +388,198 @@ impl Runtime {
-        compute_usage = tracing::field::Empty,
-    ))]
-    fn process_transaction(
-        &self,
-        state_update: &mut TrieUpdate,
-        apply_state: &ApplyState,
-        validated_tx: &ValidatedTransaction,
-        transaction_cost: &TransactionCost,
-        stats: &mut ChunkApplyStatsV0,
-    ) -> Result<(Receipt, ExecutionOutcomeWithId), InvalidTxError> {
-        let span = tracing::Span::current();
-        metrics::TRANSACTION_PROCESSED_TOTAL.inc();
-        let (mut signer, mut access_key) = get_signer_and_access_key(state_update, &validated_tx)?;
+        let mut processed_transactions = Vec::with_capacity(batch.indices.len());
 
-        let verification_result = verify_and_charge_tx_ephemeral(
-            &apply_state.config,
-            &mut signer,
-            &mut access_key,
-            validated_tx,
-            transaction_cost,
-            Some(apply_state.block_height),
-        );
-        let verification_result = match verification_result {
-            Ok(ok) => ok,
-            Err(e) => {
-                metrics::TRANSACTION_PROCESSED_FAILED_TOTAL.inc();
-                state_update.rollback();
-                return Err(e);
+        for (idx, signed_tx) in batch.iter() {
+            let tx_hash = signed_tx.get_hash();
+
+            let validated_tx = match validate_transaction(
+                &apply_state.config,
+                signed_tx.clone(),
+                current_protocol_version,
+            ) {
+                Ok(ok) => ok,
+                Err((err, _orig)) => {
+                    processed_transactions.push(ProcessedTransaction {
+                        index: idx,
+                        result: PerTransactionResult::Failure { tx_hash, error: err },
+                    });
+                    continue;
+                }
+            };
+
+            let cost = match tx_cost(&apply_state.config, validated_tx.to_tx(), gas_price) {
+                Ok(c) => c,
+                Err(e) => {
+                    processed_transactions.push(ProcessedTransaction {
+                        index: idx,
+                        result: PerTransactionResult::Failure { tx_hash, error: e.into() },
+                    });
+                    continue;
+                }
+            };
+
+            if ephemeral_signer.is_none() {
+                match get_signer_and_access_key(state_update, &validated_tx) {
+                    Ok((signer_acct, ak)) => {
+                        ephemeral_signer = Some(signer_acct);
+                        ephemeral_keys.insert(validated_tx.public_key().clone(), ak);
+                    }
+                    Err(e) => {
+                        processed_transactions.push(ProcessedTransaction {
+                            index: idx,
+                            result: PerTransactionResult::Failure { tx_hash, error: e },
+                        });
+                        continue;
+                    }
+                }
```

Comment:
Observation: it seems weird to me that we're batching up the read for account and access_key here, especially since we kinda go through checking the access key later and reading it out (potentially) later anyway.

I'd say this could be simplified quite a bit by ~inlining `get_signer_and_access` key and then deduplicating the logic with the two blocks below.

WDYT?